src/qtics/experiment.py

```diff
@@ -0,0 +1,206 @@
+                    if isinstance(item, h5py.Dataset):
+                        data[key] = np.asarray(group.get(key))
+                    elif isinstance(item, h5py.Group):
+                        data[key] = _recurse(h5file, path + key + "/")
+            return data
+
+        if not data_file:
+            data_file = self.data_file
+
+        with h5py.File(data_file, "r") as h5file:
+            return _recurse(h5file, "/")
+
+    def save_config(self):
+        """Save experiment's attributes and instruments defaults."""
+        config_attr = {
+            key: getattr(self, key)
+            for key in dir(self)
+            if isinstance(getattr(self, key), (int, float, str, bool))
+            and not key.startswith("_")
+        }
+        self.append_data_group("config", **config_attr)
+        for inst_name in self.inst_names:
+            if hasattr(self, inst_name):
+                inst = getattr(self, inst_name)
+                self.append_data_group(inst_name, parent_name="config", **inst.defaults)
+
+
+class MonitorExperiment(BaseExperiment):
+    """Base monitoring experiment class."""
+
+    sleep: float = 5
+
+    def watch(self, event: Event):
+        """Run the experiment continuously until event is set."""
+        self.all_instruments("connect")
+        log.info("Running monitor %s", self.name)
+        while True:
+            time.sleep(self.sleep)
+            self.main()
+            if event.is_set():
+                log.info("Trigger event set, %s shutting down.", self.name)
+                return
+
+
+class Experiment(BaseExperiment):
+    """Experiment with monitoring functions."""
+
+    def __init__(self, name, data_file="", data_dir="data"):
+        """Initialize."""
+        super().__init__(name, data_file=data_file, data_dir=data_dir)
+        self.monitors = []
+        self.event = Event()
+
+    def add_monitor(self, monitor: MonitorExperiment):
+        """Add monitoring experiment."""
+        self.monitors.append(monitor)
+
+    def run(self):
+        """Run the experiment with parallel monitoring functions."""
+        self.event.clear()
+        if len(self.monitors) == 0:
+            super().run()
+        else:
+            with ThreadPoolExecutor(1 + len(self.monitors)) as executor:
+                futures = []
+                log.info("Starting experiment %s and monitors.", self.name)
+                for m in self.monitors:
+                    futures.append(executor.submit(m.watch, self.event))
+                futures.append(executor.submit(self.main))
+                done, _ = wait(futures, return_when=FIRST_COMPLETED)
+                if len(done) > 0 and len(done) != len(futures):
+                    future = done.pop()
+                    if future.exception() is not None:
+                        log.warning(
+                            "One task failed with: %s, shutting down.",
+                            future.exception(),
+                        )
```

Comment:
In the case of exceptions the return is automatic? 
If the return is usable, I would also consider it for another de-nesting